### DISCUSSION TITLE
Try / except blocks not working (maybe because syncified?)

### DISCUSSION THREAD

**Discussion by Fragaile** [kaliiiiiiiiii/Selenium-Driverless#80] [General]
I want to check the title of an element inside an iframe continuously until the title reads a specific string like "Enter code" but I'm pretty sure due to the syncified nature of the library try / except blocks don't work like I expect them to. Do you have a way to do this differently or can maybe tell me what I'm doing wrong?

Expected behaviour - prints title when found.
Issue - doesn't print anything, when putting the print statement in the loop, also doesn't print anything.

Code:
```
while True:
    try:
        iframe = driver.find_element(By.TAG_NAME, "iframe")
        security = iframe.content_document
        title = security.find_element(By.XPATH, '/html/head').get_attribute('title')
        if title == 'Enter code':
            break
    except:
        pass
print(title)
```

P.S. Sorry for asking for so much help, I always try to figure this stuff out but there's very limited documentation. Thank you for the help :)

**Comment by ttlns** [comment]
@Fragaile Did you check if `security.find_element(By.XPATH, '/html/head').get_attribute('title')` actually returns the actual title when loaded?

If it doesn't, maybe try: `elem.execute_script("return obj['title']")`

Also, I can't test that unless you provide the url

The sync version technically shouldn't break any `try//except` stuff

**Reply by Fragaile** [reply]
Yes, I did, it works completely fine when I sleep for 5 seconds to let the element load in, but when I use this approach (to detect the element as soon as possible) it doesn't print anything at all. I can't really provide a url because the site I'm testing on requires you to be logged in.. Maybe I can find a similar site which has an iframe that takes a couple of seconds to load in.. Or maybe another example, one sec.

**Reply by ttlns** [reply]
yep, just give me a reproducible example and I'll test it

**Reply by Fragaile** [reply]
You're right, the problem is probably on my end since this is a very weird edge case. If I figure it out I'll post the solution. I can't replicate this behavior on any other site I've tried.

**Reply by Fragaile** [reply]
The problem I'm facing is that this line ``security = iframes[0].content_document`` hangs and the code doesn't continue..

**Reply by Fragaile** [reply]
I can't really see what's wrong since my IDE doesn't even recognise ``.content_document``

![image](https://github.com/kaliiiiiiiiii/Selenium-Driverless/assets/58077420/730f8580-96d5-48f3-a5de-266859825098)

Sometimes it works but it's hit or miss and I can't check what ``.content_document`` does.

**Reply by Fragaile** [reply]
Damn.. turns out all I had to do was sleep for 0.1 right before ``iframes[0].content_document``

**Reply by ttlns** [reply]
yep, that's because of the wait the syncification works. Currently, the IDE's thinks that `driver.find_element(By.TAG_NAME, "iframe")` returns a `asyncio coroutine` (would have to get called with `await` to get the result).

At runtime with a debugger, it should work tho

**Comment by ttlns** [comment]
> Damn.. turns out all I had to do was sleep for 0.1 right before `iframes[0].content_document`

Uh that's bad. Had that issue before as well.

https://github.com/kaliiiiiiiiii/Selenium-Driverless/blob/232adb7260e4bd3118d09ca3b987d1ae8dcc120b/src/selenium_driverless/types/webelement.py#L193

theoretically should wait for the iframe to get loaded (iframes always seem to start at `about:blank`).

@Fragaile 
Can you print what:
`elem.document_url` and `elem.source` is before those `driver.sleep(0.1)` ?

**Reply by Fragaile** [reply]
@kaliiiiiiiiii 
Before the sleep:
``elem.document_url`` is ``None``
``elem.source`` is ``<iframe src="https://checkoutshopper-live.adyen.com/checkoutshopper/securedfields/live_XXXXXXXXX/4.5.1/securedFields.html?type=card&amp;d=XXXXXXXXXXXXXXXXXXX" class="js-iframe" title="Iframe for security code" allowtransparency="true" style="border: none; height:100%; width:100%; overflow:hidden;" referrerpolicy="origin"><p>Your browser does not support iframes.</p></iframe>``

I changed some details in the url to X just to be safe.

**Reply by ttlns** [reply]
Surely, thanks! Seems to me like it still didn't load just yet fully. Wonder what the reason for that is.

**Reply by Fragaile** [reply]
How can I detect that it has fully loaded in? Else my entire script breaks and I need it to be as fast as possible so I don't want to use sleep..

**Reply by ttlns** [reply]
Maybe you could do a loop until `iframe_document.document_url` has the desired url.

**Reply by ttlns** [reply]
or use:
```python
async for data in await driver.get_cdp_event_iter("Page.frameNavigated"):
                        frame = data["frame"]
                        if frame["id"] == await iframe.__frame_id__:
                            break
```
Note: sync not supported

**Reply by Fragaile** [reply]
``.document_url`` always returns ``None``

**Reply by ttlns** [reply]
That's weird. What's about `elem.execute_script("return document.location.href")` ?

**Reply by Fragaile** [reply]
That does return the URL of the site I'm on.. but how does that help? When it returns the site URL, the iframe is fully loaded?

**Reply by ttlns** [reply]
Maybe, you can try if that's the case:)

**Comment by Fragaile** [comment]
> Note: sync not supported

Damn.. :/

**Reply by Fragaile** [reply]
Guess I'll stick to sleep for now but I hope there's a better way with sync in the future..

**Reply by Fragaile** [reply]
Sleep is still very unreliable, let me know if you decide to look into this. In the meantime I'll be going back to my old slower but reliable code unfortunately :/
